### PR TITLE
[All] OpenTracing to OpenTelemetry migration

### PR DESCRIPTION
**Pull request by SaveTheRbtz**
Issue: onflow/flow-go#2796

OpenTracing spec is deprecated: opentracing/specification#163.  Switching to the supported OpenTelemetry implementation.

Note that new tracing code would honor [OTEL env vars](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.12.0/specification/protocol/exporter.md) instead of JAEGER ones, e.g.:
`JAEGER_ENDPOINT` -> `OTEL_EXPORTER_OTLP_TRACES_ENDPOINT`, `OTEL_EXPORTER_OTLP_TRACES_INSECURE`, etc

(cc: @sjonpaulbrown @haroldsphinx)

New API is slightly slower and consumes more memory but does substantially less allocs (~ -50%):
```
name                              old time/op    new time/op    delta
StartSpanFromParent-8                485ns ± 6%     639ns ± 1%  +31.71%  (p=0.000 n=10+9)
StartTransactionSpan/cacheHit-8      993ns ± 4%     927ns ± 3%   -6.69%  (p=0.000 n=10+10)
StartTransactionSpan/cacheMiss-8    2.24µs ± 1%    2.19µs ± 1%   -2.42%  (p=0.000 n=10+10)

name                              old alloc/op   new alloc/op   delta
StartSpanFromParent-8                 577B ± 0%      864B ± 0%  +49.74%  (p=0.000 n=10+10)
StartTransactionSpan/cacheHit-8       776B ± 1%      944B ± 0%  +21.65%  (p=0.000 n=10+8)
StartTransactionSpan/cacheMiss-8    2.16kB ± 1%    2.39kB ± 0%  +10.62%  (p=0.000 n=10+10)

name                              old allocs/op  new allocs/op  delta
StartSpanFromParent-8                 7.00 ± 0%      4.00 ± 0%  -42.86%  (p=0.000 n=10+10)
StartTransactionSpan/cacheHit-8       10.3 ± 7%       6.0 ± 0%  -41.75%  (p=0.000 n=10+10)
StartTransactionSpan/cacheMiss-8      35.3 ± 2%      18.0 ± 0%  -49.01%  (p=0.000 n=10+10)
```

An example of both extensive tracing and cadence tracing enabled:
Screen Shot 2022-07-25 at 10 21 07 PM

- [x] test against localnet's Tempo instance.
- [x] noop tracer.
- [x] log tracer.
- [x] merge cadence bits onflow/cadence#1824.
- [x] merge onflow/flow-core-contracts#298.
- [x] update cadence.

### CODE REVIEW

**Review by pattyshack — COMMENTED** [review]
it's hard to review large sweeping changes to both api and implementation at the same time.  next time use the bridge pattern to decouple the changes

File: module/trace/trace.go [hunk 1]
```diff
@@ -112,141 +132,131 @@ func (t *OpenTracer) Done() <-chan struct{} {
...
+	parentSpan trace.Span,
 	operationName SpanName,
-	opts ...opentracing.StartSpanOption,
-) opentracing.Span {
-	if _, ok := span.(*NoopSpan); ok {
-		return &NoopSpan{&NoopTracer{}}
+	opts ...trace.SpanStartOption,
+) trace.Span {
+	if !parentSpan.SpanContext().IsSampled() {
+		return t.noopSpan
```
Comment (ramtinms) [inline module/trace/trace.go:223]: thanks for reusing noop

File: module/trace/trace.go [hunk 2]
```diff
@@ -112,141 +132,131 @@ func (t *OpenTracer) Done() <-chan struct{} {
...
 	}
-	opts = append(opts, opentracing.ChildOf(span.Context()))
-	return t.Tracer.StartSpan(string(operationName), opts...)
+	ctx := trace.ContextWithSpan(context.Background(), parentSpan)
+	_, span := t.tracer.Start(ctx, string(operationName), opts...)
+	return span
```
Comment (ramtinms) [inline module/trace/trace.go:227]: is `ContextWithSpan` automatically label the span relationship as children/parent ?

File: module/trace/trace.go [hunk 3]
```diff
@@ -112,141 +132,131 @@ func (t *OpenTracer) Done() <-chan struct{} {
...
 	end := time.Now()
 	start := end.Add(-duration)
-	opts = append(opts, opentracing.FollowsFrom(span.Context()))
-	opts = append(opts, opentracing.StartTime(start))
-	sp := t.Tracer.StartSpan(string(operationName), opts...)
-	sp.FinishWithOptions(opentracing.FinishOptions{FinishTime: end, LogRecords: logs})
+	ctx := trace.ContextWithSpanContext(context.Background(), parentSpan.SpanContext())
+	opts = append(opts,
+		trace.WithAttributes(attrs...),
+		trace.WithTimestamp(start),
```
Comment (ramtinms) [inline module/trace/trace.go:245]: same here how do we label it as FollowsFrom label.

File: engine/execution/computation/computer/computer.go [hunk 4]
```diff
@@ -331,20 +330,20 @@ func (e *blockComputer) executeTransaction(
...
-		if sc, ok := txInternalSpan.Context().(jaeger.SpanContext); ok {
-			traceID = sc.TraceID().String()
-		}
+		txInternalSpan.SetAttributes(attribute.String("tx_id", txID.String()))
+		traceID = txInternalSpan.SpanContext().TraceID().String()
 	}
-	defer txInternalSpan.Finish()
+	defer txInternalSpan.End()
 
 	e.log.Info().
```
Comment (ramtinms) [inline engine/execution/computation/computer/computer.go:348]: not related to this PR, but not sure when trace_id has been removed accidentally from here, including them inside the logs helps the Grafana cloud to connect logs and traces. Maybe we could put it back.

**Review by ramtinms — APPROVED** [review]
Looks good to me, just some minor questions

**Review by pattyshack — APPROVED** [review]

File: engine/execution/computation/computer/computer.go [hunk 5]
```diff
@@ -331,20 +330,20 @@ func (e *blockComputer) executeTransaction(
...
-	txSpan.LogFields(log.Int("col_index", collectionIndex))
-	defer txSpan.Finish()
+	txSpan.SetAttributes(
+		attribute.String("tx_id", txID.String()),
+		attribute.Int64("tx_index", int64(txIndex)),
+		attribute.Int("col_index", collectionIndex),
+	)
+	defer txSpan.End()
 
 	var traceID string
```
[the remark this answers is not in the document]
Comment (pattyshack) [reply engine/execution/computation/computer/computer.go:340]: logging an uninitialized traceId is pointless.  just rm it from line 403 and rm the decl here (same with my other comment)

File: module/trace/trace.go (hunk 3 above)
Comment (SaveTheRbtz) [reply module/trace/trace.go:246]: do we need to add this label? IIUC there should be no need for additional labels/links given that we have parent-child relationship here.

File: engine/execution/computation/computer/computer.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (SaveTheRbtz) [reply engine/execution/computation/computer/computer.go:340]: I think the goal here is twofold:
* for traces that are sampled we want to log traceid, but for ones that are not we do not want to log 000000000000...
* for traces that are not sampled, logging "" is cheaper than creating a whole new logger within an `if isSampled {}` block.

File: module/trace/trace.go (hunk 2 above)
Comment (SaveTheRbtz) [reply module/trace/trace.go:224]: yep!

**Comment by SaveTheRbtz** [comment]
bors merge

**Pull request merged**